File: clojure/cubesat-backend/src/cubesat_clj/telemetry/cubesat_telemetry.clj
(ns cubesat-clj.telemetry.cubesat-telemetry
  "Specifies/parses formats used by the RockBlock API and cubesat.
  These formats are described in the RockBlock web services docs at:
  https://docs.rock7.com/reference#push-api
  And for the satellite in the Google Drive for Alpha"
  (:require [clojure.string :as str]
            [cubesat-clj.config :as cfg]
            [cubesat-clj.databases.elasticsearch :as es]
            [cubesat-clj.databases.image-database :as img]
            [cubesat-clj.telemetry.rockblock-telemetry :as rb]
            [cubesat-clj.util.binary.binary-reader :as reader]
            [cubesat-clj.util.binary.hex-string :as hex]))

(def ^:const opcodes
  "Packet opcodes for cubesat (see Alpha documentation for specification)"
  {99  ::normal-report
   24  ::deployment-report
   42  ::ttl})

(defn save-cubesat-data
  "Saves a cubesat report to elasticsearch"
  [data]
  (println "cs: save-cubesat-data")
  (let [index (cfg/cubesat-db-index (cfg/get-config))]
    (es/index! index es/daily-index-strategy data)))

(defn- map-range
  "Recreation of Arduino map() function used in flight code in order to convert imu data to correct imu values.
  https://www.arduino.cc/reference/en/language/functions/math/map/"
  [x in-min in-max out-min out-max]
  (println "cs: map-range")
  (+ out-min
     (* (/ (- out-max out-min)
           (- in-max in-min))
        (- x in-min))))

;;================================================================================================
;; fragment-number is the number of the fragment
;; fragment-list is a list of different fragment numbers received
;; fragment-data is the Hex String downlinked from the IMU. Includes fragment number, x-gyro values, y-gyro values, z-gyro values
(def fragment-list (atom {:list []}))

(def imu-display-info (atom {:latest-fragment 0
                             :missing-fragments "None"
                             :highest-fragment 0}))

(defn- generate-missing-fragments 
  "Finds the missing fragments and the highest fragment received for IMU deployment downlinks. 
  It counts through all already-received fragments every time because fragments could be receieived in random order."
  [frag-list]
  (println "cs: generate-missing-fragments")
  (let [max-frag (apply max frag-list)]
    (swap! imu-display-info assoc :missing-fragments "")
    (swap! imu-display-info assoc :highest-fragment max-frag)
    (loop [x 0]
      (when (< x max-frag)
        (if (some #(= x %) frag-list)
          ()
          (let [missing-frags (:missing-fragments @imu-display-info)]
            (swap! imu-display-info assoc :missing-fragments (str missing-frags x " "))))
        (recur (inc x))))
    (if (= (count (:missing-fragments @imu-display-info)) 0) (swap! imu-display-info assoc :missing-fragments "None") ())))

(defn- compute-cycle-values    
  [{:keys [x-gyro y-gyro z-gyro] :as cycle-data}]
  (println "cs: compute-cycle-values")
  (assoc cycle-data
         :x-gyro (map-range (float x-gyro) 0 255 -180 180)
         :y-gyro (map-range (float y-gyro) 0 255 -180 180)
         :z-gyro (map-range (float z-gyro) 0 255 -180 180)))

(defn save-cycle-report 
  [cycle-data]
  (println "cs: save-cycle-report")
  (let [index (cfg/cycle-db-index (cfg/get-config))]
    (es/index! index es/daily-index-strategy cycle-data)))

(defn- separate-cycles
  [fragment-number fragment-data]
  (println "cs: seperate-cycles")        
  (let [frag-size (count fragment-data)]
      (loop [x 0]
        (when (< x frag-size)
          (let [second-index (+ x 2)
                third-index (+ x 4)
                fourth-index (+ x 6)
                x-gyro (Integer/parseInt (subs fragment-data x second-index) 16)
                y-gyro (Integer/parseInt (subs fragment-data second-index third-index) 16)
                z-gyro (Integer/parseInt (subs fragment-data third-index fourth-index) 16)
                raw-cycle-values {:x-gyro x-gyro, :y-gyro y-gyro, :z-gyro z-gyro}
                cycle-count (+ (* fragment-number 22) (/ x 6))] ;; every full fragment has 22 cycles, new cycle occurs every six digits in the hex string
            
                (save-cycle-report (merge (compute-cycle-values raw-cycle-values) {:cycle-count cycle-count})))
          (recur (+ x 6))))))

(defn save-deploy-report
  [data]
  (println "cs: save-deploy-report")
  (let [index (cfg/deploy-db-index (cfg/get-config))]
    (es/index! index es/daily-index-strategy data)))

(defn process-deploy-data
  "Saves a deployment report to elasticsearch"
  [data]
  (println "cs: process-deploy-data")
  (let [{:keys [imei transmit-time fragment-number fragment-data]} data
        meta {:imei imei :transmit-time transmit-time}
        newlist (conj (:list @fragment-list) fragment-number)]
    (swap! fragment-list assoc :list newlist) ;; add fragment to list
    (generate-missing-fragments (:list @fragment-list))
    (separate-cycles fragment-number fragment-data)
    (save-deploy-report (merge meta @imu-display-info))))

;;=================================================================================================

(defn save-image-data
  "Saves an image fragment report to elasticsearch"
  [data]
  (println "cs: save-image-data")
  (let [index (cfg/image-db-index (cfg/get-config))]
    (es/index! index es/daily-index-strategy data)))

(defn process-ttl-data
  "Process image fragment data sent by cubesat. Image comes over several fragments as
  rockblock only supports so much protocol. Image 'fragments' are then assembled into full images
  when fully collected, and saved into the image database"
  [ttl-data]
  (println "cs: process-ttl-data")
  (let [{:keys [imei transmit-time serial-number fragment-number fragment-data max-fragments]} ttl-data
        meta {:imei imei :transmit-time transmit-time}]
    (img/save-fragment serial-number fragment-number fragment-data)
    (img/try-save-image serial-number max-fragments)
    (let [data (img/get-img-display-info)]
      (save-image-data (merge meta data)))))

(defn- read-opcode
  "Reads the opcode of an incoming packet. If empty packet is received, returns ::empty-packet instead"
  [packet]
  (println "cs: read-opcode")
  (if (= (reader/remaining packet) 0)
    ::empty-packet
    (-> packet
        (reader/read-uint8)
        opcodes)))

(defn- read-img-hex-fragment
  "Reads the hexadecimal string of an image fragment to determine if the 
   fragment is the last fragment, which is indicated by the end-marker 'ffd9'. 
   Returns the serial number in decimal form, the fragment number in 
   decimal form, the max number of fragments in decimal form, and the hex string 
   of fragment data needed to be read (minus the opcode, image serial number and 
   fragment number).
   
   If the image fragment is not last, then the max number of fragments is set 
   arbitrarilty to -1, and the entirety of the fragment portion in hex string 
   must be read. If the image fragment is the last, then the max number of 
   fragments is set to (last fragment number + 1), and the hexadecimal string is 
   read up to 'ffd9'.
   
   Notes: 
          - The hex string is the data report minus the op code at the beginning
            (so everything after the op code from the downlink).
          - The serial number is stored in the indices of [0, 2) of the hex 
            string.
          - The fragment number is stored in the indices [8, 10) of the hex 
            string. Fragment count starts at 0.
          - The fragment data starts at index 10 and goes to the end of of the
            hex-string or to the end of end-marker 'ffd9'.
          - The entire hex string for an image fragment data is 69 bytes long 
            (138 characters)."
  [rockblock-data]
  (println "cs: read-img-hex-fragment")
  (let [serial-number (Integer/parseInt (subs rockblock-data 0 2) 16)
        fragment-number (Integer/parseInt (subs rockblock-data 8 10) 16)
        end-marker-index (str/index-of rockblock-data "ffd9")
        max-fragments (if (nil? end-marker-index) -1 (+ fragment-number 1))
        end-boundary (if (nil? end-marker-index) 138 (+ end-marker-index 4))
        fragment-data (hex/hex-str-to-bytes (subs rockblock-data 10 end-boundary))]
      {:serial-number serial-number, :fragment-number fragment-number, :max-fragments max-fragments, :fragment-data fragment-data}))

(defn- compute-normal-report-values
  [{:keys [burnwire-burn-time burnwire-armed-timeout-limit burnwire-attempts downlink-period x-mag y-mag z-mag x-gyro y-gyro z-gyro temp solar-current battery-voltage] :as cubesat-data}]
  (println "cs: compute-normal-report-values")
  (assoc cubesat-data
         :burnwire-burn-time (map-range (float burnwire-burn-time) 0 255 0 60000)
         :burnwire-armed-timeout-limit (map-range (float burnwire-armed-timeout-limit) 0 255 0 86400000)
         :burnwire-attempts (map-range (float burnwire-attempts) 0 255 0 10)
         :downlink-period (map-range (float downlink-period) 0 255 1000 172800000)
         :x-mag (map-range (float x-mag) 0 255 -180 180)
         :y-mag (map-range (float y-mag) 0 255 -180 180)
         :z-mag (map-range (float z-mag) 0 255 -180 180)
         :x-gyro (map-range (float x-gyro) 0 255 -180 180)
         :y-gyro (map-range (float y-gyro) 0 255 -180 180)
         :z-gyro (map-range (float z-gyro) 0 255 -180 180)
         :temp (map-range (float temp) 0 255 0 200)
         :solar-current (map-range (float solar-current) 0 255 0 500)
         :battery-voltage (map-range (float battery-voltage) 0 255 3 5)))

(defn- read-imu-hex-fragment
  [rockblock-data]
  (println "cs: read-imu-hex-fragment")
  (let [fragment-number (Integer/parseInt (subs rockblock-data 0 2) 16)
        fragment-data (subs rockblock-data 2)]
      {:fragment-number fragment-number :fragment-data fragment-data}))

(defmulti read-packet-data
  "Reads data from a packet based on opcode.
          Note: Image data is received in fragments, :data-length bytes each, which must be assembled into a full image
          after receiving all fragments. The serial number is which image is being sent, and the fragment number
          is which part of the image being sent"
  (fn [[opcode rockblock-data]] opcode))

(defmethod read-packet-data ::ttl
  [[_ rockblock-data]]
  (println "cs: read-packet-data ttl")
  (read-img-hex-fragment rockblock-data))

(defmethod read-packet-data ::normal-report
  [[_ packet]]
  (println "cs: read-packet-data normal-report")
  (-> packet
      (reader/read-structure
       [:is-photoresistor-covered ::reader/uint8
        :is-door-button-pressed ::reader/uint8
        :mission-mode ::reader/uint8
        :fire-burnwire ::reader/uint8
        :arm-burnwire ::reader/uint8
        :burnwire-burn-time ::reader/uint8
        :burnwire-armed-timeout-limit ::reader/uint8
        :burnwire-mode ::reader/uint8
        :burnwire-attempts ::reader/uint8
        :downlink-period ::reader/uint8
        :waiting-messages ::reader/uint8
        :is-command-waiting ::reader/uint8
        :x-mag ::reader/uint8
        :y-mag ::reader/uint8
        :z-mag ::reader/uint8
        :x-gyro ::reader/uint8
        :y-gyro ::reader/uint8
        :z-gyro ::reader/uint8
        :temp ::reader/uint8
        :temp-mode ::reader/uint8
        :solar-current ::reader/uint8
        :in-sun ::reader/uint8
        :acs-mode ::reader/uint8
        :battery-voltage ::reader/uint8
        :fault-mode ::reader/uint8
        :check-x-mag ::reader/uint8
        :check-y-mag ::reader/uint8
        :check-z-mag ::reader/uint8
        :check-x-gyro ::reader/uint8
        :check-y-gyro ::reader/uint8
        :check-z-gyro ::reader/uint8
        :check-temp ::reader/uint8
        :check-solar-current ::reader/uint8
        :check-battery ::reader/uint8
        :take-photo ::reader/uint8
        :camera-on ::reader/uint8])
        compute-normal-report-values))

(defmethod read-packet-data ::deployment-report
  [[_ rockblock-data]]
  (println "cs: read-packet-data deployment-report")
  (read-imu-hex-fragment rockblock-data))

(defn- report-metadata
  "Returns rockblock metadata such as transmit time from a rockblock report as a map"
  [rockblock-report]
  (println "cs: report-metadata")
  (let [{:keys [imei transmit_time]} rockblock-report]
    {:imei          imei
     :transmit-time transmit_time}))

(defn- error-data
  "Returns a map containing a cubesat report with the error opcode, raw data, and an error message"
  [rockblock-report error-msg]
  (println "cs: error-data")
  {:telemetry-report-type ::error
   :error    error-msg
   :raw-data (:data rockblock-report)})

(defn read-cubesat-data
  "Reads the cubesat data inside a rockblock report. Returns a map containing the data if read, or an error report
  if the packet is empty or some issue occurred. If the data is successfully read, the opcode is returned in the
  result map as :telemetry-report-type. Otherwise :telemetry-report-type is set to ::error"
  [rockblock-report]
  (println "cs: read-cubesat-data")
  (let [packet (rb/get-cubesat-message-binary rockblock-report)
        op (read-opcode packet)
        rockblock-data (subs (:data rockblock-report) 2)
        meta (report-metadata rockblock-report)
        result (if (= op ::empty-packet)
                 (error-data rockblock-report "empty packet")
                 (try
                    (if (= op ::normal-report)
                    (assoc (read-packet-data [op packet]) :telemetry-report-type op)
                    (assoc (read-packet-data [op rockblock-data]) :telemetry-report-type op)) ;; rockblock-data is hex string with op code removed
                   (catch Exception e (error-data rockblock-report (.getMessage e)))))]
    (merge meta result)))

File: cubesat-backend/telemetry/cubesat_telemetry.py
import traceback
from enum import Enum

import config
import databases.elastic as es
import databases.image_database as img
import util.binary.hex_string as hex
from util.binary.binary_parser import BinaryParser


class Opcodes(int, Enum):
    """Packet Opcodes for cubesat (see Alpha documentation for specification)"""

    normal_report = 99
    imu_report = 24
    camera_report = 42
    empty_packet = 0
    error = -1


# list of all imu fragment numbers received
fragment_list = []
# keeps track of various stats regarding the received imu fragments
imu_display_info = {'latest_fragment': 0,
                    'missing_fragments': [],
                    'highest_fragment': 0}


def map_range(x, in_min, in_max, out_min, out_max):
    """
    Recreation of Arduino map() function used in flight code
    https://www.arduino.cc/reference/en/language/functions/math/map/

    :param x: the number to map
    :param in_min: the lower bound of the value’s current range
    :param in_max: the upper bound of the value’s current range
    :param out_min: the lower bound of the value’s target range
    :param out_max: the upper bound of the value’s target range
    :return: the mapped value
    """
    return out_min + (((out_max - out_min) / (in_max - in_min)) * (x - in_min))


def save_cubesat_data(data: dict):
    """
    Saves a cubesat report to elasticsearch \n
    :param data: fully processed normal report
    """
    #print(data)
    es.index(config.cubesat_db_index, es.daily_index_strategy, data)


def generate_missing_fragments(frag_list: list):
    """
    Finds the missing fragments and the highest fragment received for IMU deployment downlinks.
    Counts through all already-received fragments every time because fragments could be received in random order.

    :param frag_list: list of previously received fragments
    """
    max_frag = max(frag_list)
    imu_display_info['missing_fragments'] = []
    imu_display_info['highest_fragment'] = max_frag
    for x in range(max_frag):
        if frag_list.count(x) == 0:
            imu_display_info['missing_fragments'].append(x)


def separate_cycles(fragment_number: int, fragment_data: str):
    """
    Decodes imu cycles to retrieve the x, y, and z gyro values for each and saves them to elasticsearch \n
    :param fragment_number: id number of fragment
    :param fragment_data: hex string containing the packet's imu cycle data
    """
    for x in range(0, len(fragment_data), 6):
        x_gyro = int(fragment_data[x:x + 2], 16)
        y_gyro = int(fragment_data[x + 2:x + 4], 16)
        z_gyro = int(fragment_data[x + 4:x + 6], 16)
        # every full fragment has 22 cycles, new cycle occurs every six digits in the hex string
        cycle_count = fragment_number * 22 + x / 6

        # Maps imu cycle values from the range used for transmission (0 - 255) to their actual range (-180 - 180)
        report_data = {
            'cycle_count': int(cycle_count),
            'x_gyro': float(x_gyro) / 25 - 5,
            'y_gyro': float(y_gyro) / 25 - 5,
            'z_gyro': float(z_gyro) / 25 - 5,
        }
        print('report', report_data)
        # Saves a cycle report to elasticsearch
        es.index(config.cycle_db_index, es.daily_index_strategy, report_data)


def process_save_deploy_data(data: dict):
    """
    Generates missing imu fragments, processes imu cycles, and saves imu fragment
    summary report (latest, missing, and highest received fragments) to elasticsearch \n
    :param data: dictionary with imu fragment data
    """
    fragment_list.append(data['fragment_number'])
    imu_display_info['latest_fragment'] = data['fragment_number']
    generate_missing_fragments(fragment_list)
    print(fragment_list, imu_display_info)
    separate_cycles(data['fragment_number'], data['fragment_data'])
    es.index(config.deploy_db_index, es.daily_index_strategy,
             {**report_metadata(data), **imu_display_info})


def process_save_camera_data(data: dict):
    """
    Process image fragment data sent by cubesat. Image comes over several fragments as
    rockblock only supports so much protocol. Image 'fragments' are then assembled into full images
    when fully collected, and saved into the image database. Saves image fragment
    summary report (latest, missing, and highest received fragments) to elasticsearch \n

    :param data: image fragment report
    """
    img.save_fragment(data['serial_number'], data['fragment_number'], data['fragment_data'])
    img.try_save_image(data['serial_number'], data['max_fragments'])
    es.index(config.image_db_index, es.daily_index_strategy,
             {**report_metadata(data), **img.get_img_display_info()})


def compute_normal_report_values(data: dict) -> dict:
    """
    Maps normal report values from the range used for transmission (0 - 255) to their actual range (see spec) \n
    :param data: decoded normal report
    :return: fully processed normal report
    """
    fixed_data = {
        'burnwire_burn_time': map_range(float(data['burnwire_burn_time']), 0, 255, 0, 5000),
        'burnwire_armed_timeout_limit': map_range(float(data['burnwire_armed_timeout_limit']), 0, 255, 0, 864000000),
        # 'burnwire_attempts': map_range(float(data['burnwire_attempts']), 0, 255, 0, 10),
        'downlink_period': map_range(float(data['downlink_period']), 0, 255, 60000, 172800000),
        'x_mag': map_range(float(data['x_mag']), 0, 255, -150, 150),
        'y_mag': map_range(float(data['y_mag']), 0, 255, -150, 150),
        'z_mag': map_range(float(data['z_mag']), 0, 255, -150, 150),
        'x_gyro': map_range(float(data['x_gyro']), 0, 255, -5, 5),
        'y_gyro': map_range(float(data['y_gyro']), 0, 255, -5, 5),
        'z_gyro': map_range(float(data['z_gyro']), 0, 255, -5, 5),
        'photoresistor' : map_range(float(data['photoresistor']), 0, 255, 0, 1023),
        'temp': map_range(float(data['temp']), 0, 255, -50, 200),
        # 'solar_current': map_range(float(data['solar_current']), 0, 255, 0, 500),
        'battery_voltage': map_range(float(data['battery_voltage']), 0, 255, 3, 5)
    }
    data.update(fixed_data)
    return data


def read_imu_hex_fragment(data: str) -> dict:
    """
    Separates the fragment's id number and data and returns them in a dictionary, where
    "fragment_number" is the id number of the fragment and "fragment_data" is the hex string
    downlinked from the IMU, which includes the fragment number and the x, y, and z gyro values

    :param data: hex string containing imu fragment data
    :return: a dictionary with the fragment's id number and data
    """
    return {
        'fragment_number': int(data[0:2], 16),
        'fragment_data': data[2:]
    }


def read_img_hex_fragment(data: str) -> dict:
    """
    Reads the hexadecimal string of an image fragment to determine if the
    fragment is the last fragment, which is indicated by the end-marker 'ffd9'.
    Returns the serial number in decimal form, the fragment number in
    decimal form, the max number of fragments in decimal form, and the hex string
    of fragment data needed to be read (minus the opcode, image serial number and
    fragment number).

    If the image fragment is not last, then the max number of fragments is set
    arbitrarily to -1, and the entirety of the fragment portion in hex string
    must be read. If the image fragment is the last, then the max number of
    fragments is set to (last fragment number + 1), and the hexadecimal string is
    read up to 'ffd9'.

    Notes:
          - The hex string is the data report minus the op code at the beginning
            (so everything after the op code from the downlink).
          - The serial number is stored in the indices of [0, 2) of the hex string.
          - The fragment number is stored in the indices [8, 10) of the hex string.
            Fragment count starts at 0.
          - The fragment data starts at index 10 and goes to the end of the hex string
            or to the end of end-marker 'ffd9'.
          - The entire hex string for an image fragment data is 69 bytes long (138 characters).

    :param data: hex string containing image fragment data
    :return: a dictionary with the fragment's serial #, id #, data and the total # of fragments
    """

    fragment_number = int(data[8:10], 16)
    end_marker_present = data.count('ffd9') != 0
    end_marker_index = data.index('ffd9') if end_marker_present else -1
    max_fragments = fragment_number + 1 if end_marker_present else -1
    end_boundary = end_marker_index + 4 if end_marker_present else 138
    fragment_data = hex.hex_str_to_bytes(data[10:end_boundary])
    return {
        'serial_number': int(data[0:2], 16),
        'fragment_number': fragment_number,
        'max_fragments': max_fragments,
        'fragment_data': fragment_data
    }


def report_metadata(rockblock_report: dict) -> dict:
    """
    Returns rockblock metadata such as imei and transmit time from a rockblock report as a dictionary.

    :param rockblock_report: raw data report from rockblock API
    :return: map containing the report's imei and transmit time
    """
    return {
        'imei': rockblock_report['imei'],
        'transmit_time': rockblock_report['transmit_time']
    }


def error_data(rockblock_report: dict, error_msg: str) -> dict:
    """
    Returns a map containing a cubesat report with the error opcode, raw data, and an error message.

    :param rockblock_report: raw data report from rockblock API
    :param error_msg: the error message
    :return: an error report
    """
    return {
        'telemetry_report_type': Opcodes.error,
        'error': error_msg,
        'raw_data': rockblock_report['data']
    }


def read_cubesat_data(rockblock_report: dict) -> dict:
    """
    Reads the cubesat data inside a rockblock report. Returns a map containing the data if read,
    or an error report if the packet is empty or some issue occurred. If the data is successfully read,
    the opcode is returned in the result map under the key "telemetry_report_type". Otherwise, the
    "telemetry_report_type" key is set to Opcodes.error

    :param rockblock_report: raw data report from rockblock API
    :return: processed/decoded rockblock report or error report if empty or exception occurred
    """

    # Convert hex encoded data into a python byte array
    binary_data = hex.hex_str_to_bytes(rockblock_report['data'])

    # Read opcode of report
    parser = BinaryParser(binary_data)
    if parser.remaining() == 0:
        opcode = Opcodes.empty_packet
    else:
        opcode = Opcodes(parser.read_uint8())

    # Extract data from report (strip away opcode [0:2] and command log [57:])
    data = rockblock_report['data'][2:57]

    # Reads data from a packet based on its opcode
    if opcode == Opcodes.empty_packet:
        result = error_data(rockblock_report, 'empty packet')
    else:
        try:
            if opcode == Opcodes.normal_report:
                result = compute_normal_report_values(
                    parser.read_structure(config.normal_report_structure))
            elif opcode == Opcodes.imu_report:
                result = read_imu_hex_fragment(data)
            else:  # opcode == camera_report
                result = read_img_hex_fragment(data)
            result['telemetry_report_type'] = opcode
        except Exception:
            result = error_data(rockblock_report, traceback.format_exc())
    return {**report_metadata(rockblock_report), **result}
